src/solvers/quasi_Newton.jl

```diff
@@ -125,64 +125,65 @@ function quasi_Newton!(
             local_dir_upd; θ=cautious_function
         )
     end
-
-    p = GradientProblem(M, F, gradF; evaluation=evaluation)
-    o = QuasiNewtonOptions(
+    mgo = ManifoldGradientObjective(f, grad_f; evaluation=evaluation)
+    mp = DefaultManoptProblem(M, mgo)
+    qns = QuasiNewtonState(
         M,
-        x;
-        initial_vector=get_gradient(p, x),
+        p;
+        initial_vector=get_gradient(mp, p),
         direction_update=local_dir_upd,
         stopping_criterion=stopping_criterion,
         stepsize=stepsize,
         retraction_method=retraction_method,
         vector_transport_method=vector_transport_method,
     )
-    o = decorate_options(o; kwargs...)
-    return get_solver_return(solve(p, o))
+    qns = decorate_state(qns; kwargs...)
+    return get_solver_return(solve!(mp, qns))
 end
 
-function initialize_solver!(p::GradientProblem, o::QuasiNewtonOptions)
-    o.gradient = get_gradient(p, o.x)
-    o.sk = deepcopy(o.gradient)
-    return o.yk = deepcopy(o.gradient)
+function initialize_solver!(p::AbstractManoptProblem, s::QuasiNewtonState)
+    s.gradient = get_gradient(p, s.x)
+    s.sk = deepcopy(s.gradient)
+    return s.yk = deepcopy(s.gradient)
 end
 
-function step_solver!(p::GradientProblem, o::QuasiNewtonOptions, iter)
-    o.gradient = get_gradient(p, o.x)
-    η = o.direction_update(p, o)
-    α = o.stepsize(p, o, iter, η)
-    x_old = deepcopy(o.x)
-    retract!(p.M, o.x, o.x, α * η, o.retraction_method)
+function step_solver!(mp::AbstractManoptProblem, qns::QuasiNewtonState, iter)
+    M = get_manifold(mp)
+    qns.gradient = get_gradient(mp, qns.x)
+    η = qns.direction_update(mp, qns)
+    α = qns.stepsize(mp, qns, iter, η)
+    p_old = copy(M, get_iterate(qns))
+    retract!(M, qns.x, qns.x, α * η, qns.retraction_method)
     β = locking_condition_scale(
-        p.M, o.direction_update, x_old, α * η, o.x, o.vector_transport_method
+        M, qns.direction_update, p_old, α * η, qns.x, qns.vector_transport_method
     )
     vector_transport_to!(
-        p.M, o.sk, x_old, α * η, o.x, get_update_vector_transport(o.direction_update)
+        M, qns.sk, p_old, α * η, qns.x, get_update_vector_transport(qns.direction_update)
     )
     vector_transport_to!(
-        p.M,
-        o.gradient,
-        x_old,
-        o.gradient,
-        o.x,
-        get_update_vector_transport(o.direction_update),
+        M,
+        qns.gradient,
+        p_old,
+        qns.gradient,
+        qns.x,
+        get_update_vector_transport(qns.direction_update),
     )
-    o.yk = get_gradient(p, o.x) / β - o.gradient
-    update_hessian!(o.direction_update, p, o, x_old, iter)
-    return o
+    qns.yk = get_gradient(mp, qns.x) / β - qns.gradient
```

Comment:
Could we save this calculated gradient in `qns.gradient` at the end of this function? I'm currently seeing about 4 repeated calculation of gradient at the same point in quasi-Newton and this is one of them.

Reply:
Is it ok if we use the planned `CachedObjective` for that? Otherwise we would have t introduce a `qns.X`-or-so field.

Reply:
If there are simple algorithmic improvements (like this one) it would be a nice thing to have. At the very least we save a cache lookup.

Reply:
sure we could do that.

Reply:
where do you see 4? I see 2 and we could spare that locally with one memory allocation, sure.

Reply:
I've modified my gradient function to print hash of point every time it is evaluated. It prints the same hash value 3 or 4 times in a row, and then it changes to a different one. I haven't tracked the other ones yet, maybe they are more complicated.

Reply:
if you have debug or a stopping criterion that checks gradient change,... these evaluate as well.